File: 0437-path-sum-iii/0437-path-sum-iii.cpp
class Solution {
public:
    int ans;

    int pathSum(TreeNode* root, int targetSum) {
        if (!root) return 0;
        ans = 0;
        dfs(root, targetSum);
        return ans;
    }

    void dfs(TreeNode* root, int targetSum) {
        if (root) {
            check(root, targetSum, root->val);
            dfs(root->left, targetSum);
            dfs(root->right, targetSum);
        }
    }

    void check(TreeNode* root, int targetSum, long long sum) {
        if (root == NULL) return;
        if ((long long) targetSum == sum) ans++;
        if (root->left) check(root->left, targetSum, sum + root->left->val);
        if (root->right) check(root->right, targetSum, sum + root->right->val);
    }
};

File: 0437-path-sum-iii/0437-path-sum-iii.java
class Solution {
    private int ans;

    public int pathSum(TreeNode root, int targetSum) {
        this.ans = 0;
        dfs(root, targetSum);
        return ans;
    }

    private void dfs(TreeNode root, int targetSum) {
        if (root != null) {
            checkFromNode(root, (long) targetSum);
            dfs(root.left, targetSum);
            dfs(root.right, targetSum);
        }
    }

    private void checkFromNode(TreeNode root, long targetSum) {
        if (root == null) return;
        targetSum -= (long) root.val;
        if (targetSum == 0L) ++ans;
        if (root.left != null) checkFromNode(root.left, targetSum);
        if (root.right != null) checkFromNode(root.right, targetSum);
    }
}